vllm/v1/worker/gpu_model_runner.py

```diff
@@ -399,6 +401,21 @@ def _update_states(self, scheduler_output: "SchedulerOutput") -> None:
                     req_index, start_index:end_token_index] = spec_token_ids
             # NOTE(woosuk): `num_tokens` here may include spec decode tokens.
             self.input_batch.num_tokens[req_index] = end_token_index
+            # Fill the bitmask
+            if (req_id in scheduler_output.guided_decoding_request_ids
+                    and req_state.grammar is not None):
+                idx = scheduler_output.guided_decoding_request_ids[req_id]
+                # should already be ready
+                assert scheduler_output.grammar_bitmask is not None
+                if not req_state.grammar.prefilled:
+                    req_state.grammar.prefilled = True
+                else:
+                    if not req_state.grammar.matcher.is_terminated():
+                        # NOTE: this relies on xgrammar internal bitmask,
+                        # so we need to give the actual index
+                        # of the the request_id in the batch
+                        req_state.grammar.fill_bitmask(
+                            scheduler_output.grammar_bitmask, idx)
```

Comment:
This shouldn't happen in the model runner. The proposed design for this iteration was to advance the FSM and calculate the bitmask on the engine/scheduler side, and then broadcast that to the GPU workers.

Reply:
I think this also means we're sending the compiled grammar down to the model runner, which we also don't want